src/platform/EFR32/Efr32PsaOpaqueKeypair.cpp

```diff
@@ -0,0 +1,501 @@
+                                      size_t * output_length) const
+{
+    CHIP_ERROR error    = CHIP_NO_ERROR;
+    psa_status_t status = PSA_ERROR_BAD_STATE;
+
+    VerifyOrExit(mHasKey, error = CHIP_ERROR_INCORRECT_STATE);
+
+    status = psa_raw_key_agreement(PSA_ALG_ECDH, *(mbedtls_svc_key_id_t *) mContext, their_key, their_key_len, output, output_size,
+                                   output_length);
+
+    VerifyOrExit(status == PSA_SUCCESS, {
+        _log_PSA_error(status);
+        error = CHIP_ERROR_INTERNAL;
+    });
+
+exit:
+    return error;
+}
+
+CHIP_ERROR EFR32OpaqueKeypair::Delete()
+{
+    CHIP_ERROR error    = CHIP_NO_ERROR;
+    psa_status_t status = PSA_ERROR_BAD_STATE;
+
+    VerifyOrExit(mHasKey, error = CHIP_ERROR_INCORRECT_STATE);
+
+    status = psa_destroy_key(*(mbedtls_svc_key_id_t *) mContext);
+    VerifyOrExit(status == PSA_SUCCESS, {
+        _log_PSA_error(status);
+        error = CHIP_ERROR_INTERNAL;
+    });
+
+exit:
+    mHasKey       = false;
+    mIsPersistent = false;
+    memset(mPubkeyRef, 0, mPubkeySize);
+    if (mContext)
+    {
+        *(mbedtls_svc_key_id_t *) mContext = 0;
+    }
+    return error;
+}
+
+/*******************************************************************************
+ *
+ * PSA Crypto backed implementation of EFR32OpaqueP256Keypair
+ *
+ ******************************************************************************/
+EFR32OpaqueP256Keypair::EFR32OpaqueP256Keypair()
+{
+    mPubkeyRef    = mPubKey.Bytes();
+    mPubkeySize   = mPubKey.Length();
+    mPubkeyLength = 0;
+}
+
+EFR32OpaqueP256Keypair::~EFR32OpaqueP256Keypair() {}
+
+CHIP_ERROR EFR32OpaqueP256Keypair::Initialize()
+{
+    if (mPubkeyLength > 0)
+    {
+        // already have a key - ECDH use case where CASESession is calling Initialize()
+        return CHIP_NO_ERROR;
+    }
+
+    ChipLogError(Crypto, "Initialize() is invalid on opaque keys, use Create() instead");
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR EFR32OpaqueP256Keypair::Serialize(P256SerializedKeypair & output) const
+{
+    ChipLogError(Crypto, "Serialisation is invalid on opaque keys, share the object instead");
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR EFR32OpaqueP256Keypair::Deserialize(P256SerializedKeypair & input)
+{
+    ChipLogError(Crypto, "Serialisation is invalid on opaque keys");
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR EFR32OpaqueP256Keypair::NewCertificateSigningRequest(uint8_t * out_csr, size_t & csr_length) const
+{
+    // Copied from mbedTLS, since x.509 is part of the TLS layer.
+#if defined(MBEDTLS_X509_CSR_WRITE_C)
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    int result       = 0;
+    size_t out_length;
+
+    mbedtls_x509write_csr csr;
+    mbedtls_x509write_csr_init(&csr);
+
+    VerifyOrExit(mHasKey, error = CHIP_ERROR_INCORRECT_STATE);
+
+    psa_crypto_init();
+
+    // Use PSA key to generate CSR
```

Comment:
It should be commented/noted that mbedTLS on the platform is steering in the background, through the build of the SDK, to have the PSA implement the signing using the key passed here

Reply:
Changed implementation to use `GenerateCertificateSigningRequest` instead, which does not use mbedTLS. Thanks for the suggestion.